charpick/charpick.c

```diff
@@ -350,478 +372,550 @@ chooser_button_clicked (GtkButton *button, charpick_data *curr_data)
-    
-  if (p_curr_data->panel_vertical)
-    box = gtk_box_new (GTK_ORIENTATION_VERTICAL, 0);
-  else 
-    box = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 0);
-  gtk_widget_show (box);
-  p_curr_data->box = box;
-  
-  button = gtk_button_new ();
-  if (g_list_length (p_curr_data->chartable) != 1)
-  {
-    gtk_widget_set_tooltip_text (button, _("Available palettes"));
-  
-    switch (mate_panel_applet_get_orient (MATE_PANEL_APPLET (p_curr_data->applet))) {
-       	case MATE_PANEL_APPLET_ORIENT_DOWN:
-          	arrow = gtk_image_new_from_icon_name ("pan-down-symbolic", GTK_ICON_SIZE_MENU);
-       		break;
-       	case MATE_PANEL_APPLET_ORIENT_UP:
-          	arrow = gtk_image_new_from_icon_name ("pan-up-symbolic", GTK_ICON_SIZE_MENU);
-       		break;
-       	case MATE_PANEL_APPLET_ORIENT_LEFT:
-       		arrow = gtk_image_new_from_icon_name ("pan-start-symbolic", GTK_ICON_SIZE_MENU);
-  		break;
-       	case MATE_PANEL_APPLET_ORIENT_RIGHT:
-       		arrow = gtk_image_new_from_icon_name ("pan-end-symbolic", GTK_ICON_SIZE_MENU);
-  		break;
-    default:
-  	  g_assert_not_reached ();
+    GtkWidget *box, *button_box, **row_box;
+    GtkWidget *button, *arrow;
+    gint i = 0, len = g_utf8_strlen (p_curr_data->charlist, -1);
+    GtkWidget **toggle_button;
+    gchar *charlist;
+    gint max_width=1, max_height=1;
+    gint size_ratio;
+
+    toggle_button = g_new (GtkWidget *, len);
+
+    if (p_curr_data->box)
+        gtk_widget_destroy (p_curr_data->box);
+
+    if (p_curr_data->panel_vertical)
+        box = gtk_box_new (GTK_ORIENTATION_VERTICAL,
+                           0);
+    else 
+        box = gtk_box_new (GTK_ORIENTATION_HORIZONTAL,
+                           0);
+
+    gtk_widget_show (box);
+    p_curr_data->box = box;
+
+    button = gtk_button_new ();
+    if (g_list_length (p_curr_data->chartable) != 1)
+    {
+        gtk_widget_set_tooltip_text (button, _("Available palettes"));
+
+        switch (mate_panel_applet_get_orient (MATE_PANEL_APPLET (p_curr_data->applet))) {
+            case MATE_PANEL_APPLET_ORIENT_DOWN:
+                arrow = gtk_image_new_from_icon_name ("pan-down-symbolic",
+                                                      GTK_ICON_SIZE_MENU);
+                break;
+            case MATE_PANEL_APPLET_ORIENT_UP:
+                arrow = gtk_image_new_from_icon_name ("pan-up-symbolic",
+                                                      GTK_ICON_SIZE_MENU);
+                break;
+            case MATE_PANEL_APPLET_ORIENT_LEFT:
+                arrow = gtk_image_new_from_icon_name ("pan-start-symbolic",
+                                                      GTK_ICON_SIZE_MENU);
+                break;
+            case MATE_PANEL_APPLET_ORIENT_RIGHT:
+                arrow = gtk_image_new_from_icon_name ("pan-end-symbolic",
+                                                      GTK_ICON_SIZE_MENU);
+                break;
+        default:
+        g_assert_not_reached ();
```

Comment:
indent

Reply:
I hope it is fine now?

Reply:
It's better than before, however `default` should be aligned with `case`.

Reply:
next update, better?